File: packages/ios-client/Sources/FishjamClient/webrtc/helpers/TrackBitratesMapper.swift
import WebRTC

struct TrackBitratesMapper {
    static func mapTracksToProtoBitrates(localTracks: [Track]) -> [String: Fishjam_MediaEvents_Peer_MediaEvent
        .TrackBitrates]
    {
        Dictionary(
            uniqueKeysWithValues: localTracks.compactMap {
                t -> (String, Fishjam_MediaEvents_Peer_MediaEvent.TrackBitrates)? in
                guard let track = t as? LocalCameraTrack else { return nil }

                let bitrates: [Fishjam_MediaEvents_Peer_MediaEvent.VariantBitrate] = track.sendEncodings.compactMap {
                    param in
                    guard let ridString = param.rid else { return nil }
                    var variantBitrate = Fishjam_MediaEvents_Peer_MediaEvent.VariantBitrate()
                    let trackEncoding = try? TrackEncoding(ridString)
                    variantBitrate.variant =
                        Fishjam_MediaEvents_Variant(rawValue: trackEncoding?.rawValue ?? 0) ?? .unspecified
                    variantBitrate.bitrate = param.maxBitrateBps?.int32Value ?? 0
                    return variantBitrate
                }

                var trackBitrates = Fishjam_MediaEvents_Peer_MediaEvent.TrackBitrates()
                trackBitrates.trackID = track.webrtcId
                trackBitrates.variantBitrates = bitrates

                return (track.webrtcId, trackBitrates)
            })
    }
}

File: packages/android-client/FishjamClient/src/main/java/com/fishjamcloud/client/webrtc/helpers/TrackBitratesMapper.kt
package com.fishjamcloud.client.webrtc.helpers

import com.fishjamcloud.client.media.Track
import fishjam.media_events.Shared
import fishjam.media_events.peer.Peer

class TrackBitratesMapper {
  companion object {
    fun mapTracksToProtoBitrates(localTracks: Map<String, Track>): Map<String, Peer.MediaEvent.TrackBitrates> =
      localTracks.values.associate { track ->
        track.webrtcId() to
          Peer.MediaEvent.TrackBitrates
            .newBuilder()
            .setTrackId(track.webrtcId())
            .addAllVariantBitrates(
              track.sendEncodings.map { encoding ->
                val variant =
                  when (encoding.rid) {
                    "h" -> Shared.Variant.VARIANT_HIGH
                    "m" -> Shared.Variant.VARIANT_MEDIUM
                    "l" -> Shared.Variant.VARIANT_LOW
                    else -> Shared.Variant.VARIANT_UNSPECIFIED
                  }

                Peer.MediaEvent.VariantBitrate
                  .newBuilder()
                  .setVariant(variant)
                  .setBitrate(encoding.maxBitrateBps ?: 0)
                  .build()
              }
            ).build()
      }
  }
}
